compiler/modules/bir/verify.bal

```diff
@@ -79,6 +92,73 @@ type Error err:Semantic|err:Internal;
 function verifyBasicBlock(VerifyContext vc, BasicBlock bb) returns Error? {
     foreach Insn insn in bb.insns {
         check verifyInsn(vc, insn);
+        check verifyInsnRegisterKinds(vc, insn);
+    }
+}
+
+type MultipleOperandInsn IntBinaryInsn|IntNoPanicArithmeticBinaryInsn|FloatArithmeticBinaryInsn
+    |DecimalArithmeticBinaryInsn|CompareInsn|ListConstructInsn|ListGetInsn|MappingConstructInsn
+    |MappingGetInsn|StringConcatInsn|EqualityInsn|TypeMergeInsn;
+
+type ResultInsn readonly & record { *ResultInsnBase; };
+
+function verifyInsnRegisterKinds(VerifyContext vc, Insn insn) returns Error? {
+    if insn is ResultInsn {
+        if (vc.getNumTmpRegisters() > insn.result.number) && vc.getUsedTmpRegisters(insn.result.number) {
```

Comment:
`if tmpRegisterUsed[insn.result.number] {`

Pre-populate the tmpRegisterUsed array with `false` upto `registers.length()`. Then you don't have to check the length.